### ISSUE TITLE
[MacOS] Rendering with Docker 

### ISSUE THREAD

**Issue by vi3itor** [google-research/football#188]
Note: @shehrum posted it on the [mailing list](https://groups.google.com/forum/#!topic/google-research-football/UdIp-OdSQPg):
> I installed the docker image following the [Instructions](https://github.com/google-research/football/blob/master/gfootball/doc/docker.md) on my mac. However, I get the following errors when I try to execute the command `python3 -m gfootball.play_game --action_set=full` to play the game. Can someone guide me how to fix the issue?

> I tried to fix the issue by following this [link](https://cntnr.io/running-guis-with-docker-on-mac-os-x-a14df6a76efc).
> However, that didn't resolve the issue. 
![image](https://user-images.githubusercontent.com/3192677/85944874-f2daec80-b974-11ea-92cd-781a8388d7b9.png)

**Comment by vi3itor** [comment]
**TLDR** Looks like it is not possible, follow [the guide](https://github.com/google-research/football/#mac-os-x) for Mac OS X instead.

When I tried running: 
`docker run -e DISPLAY={MY_IP} -it -v /tmp/.X11-unix:/tmp/.X11-unix:rw gfootball bash`
`python3 -m gfootball.play_game --action_set=full`
I’ve got the same error: 
```
pygame 1.9.6
Hello from the pygame community. https://www.pygame.org/contribute.html
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default
error: XDG_RUNTIME_DIR not set in the environment.
error: XDG_RUNTIME_DIR not set in the environment.
Couldn't load GL function glBegin: Video subsystem has not been initialized
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_Z16print_stacktracev+0x25)[0x7fdc291669b5]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted16OpenGLRenderer3D21CreateContextInternalEiiibb+0x1152)[0x7fdc2929e722]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted16OpenGLRenderer3D13CreateContextEiiib+0x42)[0x7fdc292a08c2]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted14GraphicsSystem10InitializeEb+0x197)[0x7fdc29265ac7]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_Z8run_gamePN7blunted10PropertiesEb+0x7a)[0x7fdc2916deaa]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7GameEnv10start_gameEv+0x24b)[0x7fdc291b10cb]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN5boost6python7objects23caller_py_function_implINS0_6detail6callerIM7GameEnvFvvENS0_21default_call_policiesENS_3mpl7vector2IvR14GameEnv_PythonEEEEEclEP7_objectSH_+0x35)[0x7fdc29188925]
/usr/lib/x86_64-linux-gnu/libboost_python3-py36.so.1.65.1(_ZNK5boost6python7objects8function4callEP7_objectS4_+0x285)[0x7fdc28a4e855]
/usr/lib/x86_64-linux-gnu/libboost_python3-py36.so.1.65.1(+0x21a18)[0x7fdc28a4ea18]
/usr/lib/x86_64-linux-gnu/libboost_python3-py36.so.1.65.1(_ZN5boost6python21handle_exception_implENS_9function0IvEE+0x73)[0x7fdc28a553b3]
```

After fixing the `DISPLAY` argument by adding `:0` to the IP address (I should note that `socat` was running in a separate tab in the terminal) 
`docker run -e DISPLAY={MY_IP}:0 --privileged -it -v /tmp/.X11-unix:/tmp/.X11-unix:rw gfootball bash`

I got a new error:
```
root@cb04a39ba594:/gfootball# python3 -m gfootball.play_game --action_set=full
pygame 1.9.6
Hello from the pygame community. https://www.pygame.org/contribute.html
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  149 (GLX)
  Minor opcode of failed request:  6 (X_GLXIsDirect)
  Serial number of failed request:  98
  Current serial number in output stream:  97
```

I found [here](https://unix.stackexchange.com/questions/429760/opengl-rendering-with-x11-forwarding#comment823975_429760) a suggestion to enable GLX:
`defaults write org.macosforge.xquartz.X11 enable_iglx -bool true`
`glxgears` command works and I see the gears. However, they are moving really slowly.

And now a new error appears: 
```
root@eac3217608ac:/gfootball# python3 -m gfootball.play_game --action_set=full
pygame 1.9.6
Hello from the pygame community. https://www.pygame.org/contribute.html
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
Sun Jun 28 05:41:50 2020 [Warning] in [OpenGLRenderer3D::]: Failed to init OpenGL, trying with MESA.
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
shader compilation error (media/shaders/simple.vert)
Sun Jun 28 05:41:50 2020 [Warning] in [::LoadGLShader]: shader compilation error (media/shaders/simple.vert)
0 0
Sun Jun 28 05:41:50 2020 [FATAL ERROR !!! N00000 !!!] in [OpenGLRenderer3D::LoadShader]: 1Could not compile vertex program: simple
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_Z16print_stacktracev+0x25)[0x7f4d9a5719b5]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted3LogENS_9e_LogTypeENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_S6_+0x33f)[0x7f4d9a66220f]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted16OpenGLRenderer3D10LoadShaderERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_+0x1cc)[0x7f4d9a6ad90c]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted16OpenGLRenderer3D21CreateContextInternalEiiibb+0xcf2)[0x7f4d9a6a92c2]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted16OpenGLRenderer3D13CreateContextEiiib+0x1d4)[0x7f4d9a6aba54]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted14GraphicsSystem10InitializeEb+0x197)[0x7f4d9a670ac7]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_Z8run_gamePN7blunted10PropertiesEb+0x7a)[0x7f4d9a578eaa]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7GameEnv10start_gameEv+0x24b)[0x7f4d9a5bc0cb]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN5boost6python7objects23caller_py_function_implINS0_6detail6callerIM7GameEnvFvvENS0_21default_call_policiesENS_3mpl7vector2IvR14GameEnv_PythonEEEEEclEP7_objectSH_+0x35)[0x7f4d9a593925]
/usr/lib/x86_64-linux-gnu/libboost_python3-py36.so.1.65.1(_ZNK5boost6python7objects8function4callEP7_objectS4_+0x285)[0x7f4d99e59855]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_Z16print_stacktracev+0x25)[0x7f4d9a5719b5]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_Z7handleri+0x8)[0x7f4d9a5719e8]
/lib/x86_64-linux-gnu/libc.so.6(+0x3ef20)[0x7f4d9be13f20]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted3LogENS_9e_LogTypeENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_S6_+0x33f)[0x7f4d9a66220f]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted16OpenGLRenderer3D10LoadShaderERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_+0x1cc)[0x7f4d9a6ad90c]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted16OpenGLRenderer3D21CreateContextInternalEiiibb+0xcf2)[0x7f4d9a6a92c2]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted16OpenGLRenderer3D13CreateContextEiiib+0x1d4)[0x7f4d9a6aba54]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7blunted14GraphicsSystem10InitializeEb+0x197)[0x7f4d9a670ac7]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_Z8run_gamePN7blunted10PropertiesEb+0x7a)[0x7f4d9a578eaa]
/usr/local/lib/python3.6/dist-packages/gfootball_engine/_gameplayfootball.so(_ZN7GameEnv10start_gameEv+0x24b)[0x7f4d9a5bc0cb]
Error: signal 11: 
```

A closer look at the problem suggests that `OpenGL` version is too low. Running `glxinfo` confirms it: 
```
root@eac3217608ac:/gfootball# glxinfo
name of display: {MY_IP}:0
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
display: {MY_IP}:0  screen: 0
direct rendering: No (If you want to find out why, try setting LIBGL_DEBUG=verbose)
server glx vendor string: SGI
server glx version string: 1.4
server glx extensions:
    GLX_ARB_multisample, GLX_EXT_import_context, GLX_EXT_visual_info,
    GLX_EXT_visual_rating, GLX_OML_swap_method, GLX_SGIS_multisample,
    GLX_SGIX_fbconfig
client glx vendor string: Mesa Project and SGI
client glx version string: 1.4
client glx extensions:
    GLX_ARB_context_flush_control, GLX_ARB_create_context,
    GLX_ARB_create_context_no_error, GLX_ARB_create_context_profile,
    GLX_ARB_create_context_robustness, GLX_ARB_fbconfig_float,
    GLX_ARB_framebuffer_sRGB, GLX_ARB_get_proc_address, GLX_ARB_multisample,
    GLX_EXT_buffer_age, GLX_EXT_create_context_es2_profile,
    GLX_EXT_create_context_es_profile, GLX_EXT_fbconfig_packed_float,
    GLX_EXT_framebuffer_sRGB, GLX_EXT_import_context,
    GLX_EXT_texture_from_pixmap, GLX_EXT_visual_info, GLX_EXT_visual_rating,
    GLX_INTEL_swap_event, GLX_MESA_copy_sub_buffer,
    GLX_MESA_multithread_makecurrent, GLX_MESA_query_renderer,
    GLX_MESA_swap_control, GLX_OML_swap_method, GLX_OML_sync_control,
    GLX_SGIS_multisample, GLX_SGIX_fbconfig, GLX_SGIX_pbuffer,
    GLX_SGIX_visual_select_group, GLX_SGI_make_current_read,
    GLX_SGI_swap_control, GLX_SGI_video_sync
GLX version: 1.4
GLX extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, GLX_EXT_import_context,
    GLX_EXT_visual_info, GLX_EXT_visual_rating, GLX_OML_swap_method,
    GLX_SGIS_multisample, GLX_SGIX_fbconfig, GLX_SGIX_pbuffer,
    GLX_SGI_make_current_read
OpenGL vendor string: Intel Inc.
OpenGL renderer string: Intel(R) Iris(TM) Graphics 550
OpenGL version string: 1.4 (2.1 INTEL-12.10.17)
OpenGL extensions:
	...(list of extensions continues)
```

**Bottom line**: 
I’m not an expert here, but I think it might be very difficult or even impossible to run gfootball or any other decent graphic application in Docker on macOS. As explained in [this comment](https://bugs.freedesktop.org/show_bug.cgi?id=93953#c8), Apple is no longer supporting XQuartz, neither other developers. The latest version was released in 2016. 

You can try investigating further, and please share your findings. 
But if you just want to run the game with rendering, follow [the guide](https://github.com/google-research/football/#mac-os-x) for Mac OS X. It works. And if there are any problems, feel free to ask for solutions here or open a new issue.

**Comment by shehrum** [comment]
So I found out that GRF doesn't support rendering in Docker. I chose to go with Docker as I was unable to follow the Mac OS X guide for installation, as it gave me a lot of installation issues. I will try again on that path, and share issues, and hopefully we could diagnose, fix and share. Thanks!

**Comment by shehrum** [comment]
I'm having a hard time fixing the python boost issue on my mac. When I run the command `pip3 install .` I get the following error. 

`    Using Python: 3.7.7
    -- Could NOT find Boost: missing: python3-py37 (found /usr/local/lib/cmake/Boost-1.72.0/BoostConfig.cmake (found version "1.72.0"))
    python3-py37 not found. Trying other names.
    -- Could NOT find Boost: missing: python-py37 (found /usr/local/lib/cmake/Boost-1.72.0/BoostConfig.cmake (found version "1.72.0"))
    python-py37 not found. Trying other names.
    -- Could NOT find Boost: missing: python37 (found /usr/local/lib/cmake/Boost-1.72.0/BoostConfig.cmake (found version "1.72.0"))
    CMake Error at CMakeLists.txt:71 (message):
      Python boost not found`
    



I saw someone with the same issue [here](https://github.com/google-research/football/issues/51), but the comments didn't help me get anywhere yet. Any pointers will be greatly appreciated.


I am using `Mac OS`, `Python 3.7.7`, and have `boost` & `boost-python3` installed with brew.

**Comment by shehrum** [comment]
Issue fixed by running brew ` unlink boost-python3` , as I had boost 1.72.0_1 installed, and had to revert it to 1.72.0 to make it work.

**Comment by vi3itor** [comment]
Good to hear that you’ve found a fix for your set up and shared it here. 
Unlike as with `pip3`, it is quite difficult to install a specific version of package in `brew`. And since `boost` and `boost-python` are getting regular upgrades and moving to python 3.8, we should probably use a specific earlier version (formula) for `boost`, same as we do for `boost-python`. 
The other idea could be trying out Anaconda, because it ships with a compatible version of boost-python (probably it has boost too), but there were some other issues reported. If anyone uses Anaconda and can test it and prepare a guide, would be awesome.
